File: Arrays/readability/readability.c
#include <stdio.h>
#include <cs50.h>
#include <ctype.h>
#include <math.h>
#include <string.h>

// hints the compiler that these functions are gonna be implemented later
string get_text(void);
int calculate_grade(int letters, int words, int sentences);

int main(void)
{
    // declare variables
    string text = get_text();
    int letters = 0;
    // words declared with 1 because otherwise one word will not be counted,
    // due to its count depend on the space character
    int words = 1;
    int sentences = 0;

    // check every character
    for (int i = 0, n = strlen(text); i < n; i++)
    {
        // if the character is alphabetic, increase letter variable
        if (isalpha(text[i]))
        {
            letters++;
        }
        // if the character is a space (or something like), increase words variable
        else if (isspace(text[i]))
        {
            words++;
        }
        // if the character is a punctuation, increase sentences variable
        else if (ispunct(text[i])
                 && (text[i] == '.'
                     || text[i] == '!'
                     || text[i] == '?'))
        {
            sentences++;
        }
    }

    // calculates the grade
    int grade = calculate_grade(letters, words, sentences);

    if (grade < 1)
    {
        printf("Before Grade 1\n");
    }
    else if (grade >= 1 && grade <= 16)
    {
        printf("Grade %i\n", grade);
    }
    else
    {
        printf("Grade 16+\n");
    }

    return 0;
}

// gets the text from the user
string get_text(void)
{
    string text = get_string("Text: ");

    return text;
}

// calculates the grade according to the number of letters, words and sentences
int calculate_grade(int letters, int words, int sentences)
{
    float hundred_words = words / 100.0;
    float l = letters / hundred_words;
    float s = sentences / hundred_words;

    float grade = 0.0588 * l - 0.296 * s - 15.8;

    return round(grade);
}

File: Python/readability/readability.py
# calculates the correspondent grade of a given text, using the Coleman-Liau
# index
def main():
    text = input('Text: ')

    letters, words, sentences = count_letters_words_sentences(text)
    grade = calculate_grade(letters, words, sentences)

    if (grade < 1):
        print('Before Grade 1')

    elif (grade < 16):
        print(f'Grade {grade}')

    else:
        print('Grade 16+')


# counts the letters, words and sentences of a given text
def count_letters_words_sentences(text: str) -> tuple:
    letters = 0
    words = 1
    sentences = 0

    for c in range(len(text)):

        if (text[c].isalpha()):
            letters += 1

        elif (text[c].isspace()):
            words += 1

        # valid punctuations
        elif (text[c] == '.' or text[c] == '!' or text[c] == '?'):
            sentences += 1

    return letters, words, sentences


# using the Coleman-Liau index, calculates the grade of a text using the words per hundred words
# and senteces per hundred words
def calculate_grade(letters: int, words: int, sentences: int) -> int:
    hundred_words = words / 100
    l = letters / hundred_words
    s = sentences / hundred_words

    grade = .0588 * l - .296 * s - 15.8

    return round(grade)


main()